itest/lnd_payment_test.go

```diff
@@ -767,3 +770,128 @@ func assertChannelState(ht *lntest.HarnessTest, hn *node.HarnessNode,
+
+	// Connect the interceptor.
+	interceptor, cancelInterceptor := bob.RPC.HtlcInterceptor()
+
+	// Prepare the test cases.
+	addResponse := carol.RPC.AddInvoice(&lnrpc.Invoice{
+		ValueMsat: 1000,
+	})
+	invoice := carol.RPC.LookupInvoice(addResponse.RHash)
+
+	// We initiate a payment from Alice and define the payment context
+	// cancellable.
+	done := make(chan struct{})
+	ctx, cancelPaymentContext := context.WithCancel(context.Background())
+	var paymentStream rpc.PaymentClient
+	go func() {
+		// Signal that all the payments have been sent.
+		defer close(done)
+
+		req := &routerrpc.SendPaymentRequest{
+			PaymentRequest: invoice.PaymentRequest,
+			TimeoutSeconds: 60,
+			FeeLimitSat:    100000,
+			Cancelable:     true,
+		}
+
+		paymentStream = alice.RPC.SendPaymentWithContext(ctx, req)
+	}()
+
+	// We start the htlc interceptor with a simple implementation that
+	// saves all intercepted packets. These packets are held to simulate a
+	// pending payment.
+	packet := ht.ReceiveHtlcInterceptor(interceptor)
+
+	// Here we should wait for the channel to contain a pending htlc, and
+	// also be shown as being active.
+	err := wait.NoError(func() error {
+		channel := ht.QueryChannelByChanPoint(bob, cpAB)
+
+		if len(channel.PendingHtlcs) == 0 {
+			return fmt.Errorf("expect alice <> bob channel to " +
+				"have pending htlcs")
+		}
+		if channel.Active {
+			return nil
+		}
+
+		return fmt.Errorf("channel not active")
+	}, defaultTimeout)
+	require.NoError(
+		ht, err, "alice <> bob channel pending htlc never arrived",
+	)
+
+	// Ensure that Alice's payment is in-flight because Bob is holding the
+	// htlc.
+	ht.AssertPaymentStatusFromStream(paymentStream, lnrpc.Payment_IN_FLIGHT)
+
+	// Cancel the payment context. This should end the payment stream
+	// context, but the payment should still be in state in-flight without a
+	// failure reason.
+	cancelPaymentContext()
+
+	var preimage lntypes.Preimage
+	copy(preimage[:], invoice.RPreimage)
+	ht.AssertPaymentStatus(alice, preimage, lnrpc.Payment_IN_FLIGHT)
+	reasonNone := lnrpc.PaymentFailureReason_FAILURE_REASON_NONE
+	ht.AssertPaymentFailureReason(alice, preimage, reasonNone)
+
+	// Bob fails the payment which fails Alice's payment.
+	err = interceptor.Send(&routerrpc.ForwardHtlcInterceptResponse{
+		IncomingCircuitKey: packet.IncomingCircuitKey,
+		Action:             routerrpc.ResolveHoldForwardAction_FAIL,
+	})
+	require.NoError(ht, err, "failed to send request")
+	ht.AssertPaymentStatus(alice, preimage, lnrpc.Payment_FAILED)
+
+	// Since the payment context was cancelled, no further payment attempts
+	// should've been made, and we observe FAILURE_REASON_CANCELED.
+	reasonCanceled := lnrpc.PaymentFailureReason_FAILURE_REASON_CANCELED
+	ht.AssertPaymentFailureReason(alice, preimage, reasonCanceled)
+
+	// Cancel the context, which will disconnect the above interceptor.
+	cancelInterceptor()
+
+	// Make sure all goroutines are finished.
+	select {
```

Comment:
this is unnecessary as we know `paymentStream` was returned in the above goroutine.